File: Filling Bookcase Shelves.py
class Solution:
    def minHeightShelves(self, books: List[List[int]], shelfWidth: int) -> int:
        n = len(books)
        dp = {}

        def place(book, j, h):
            # print(book, j, h)
            nonlocal n, dp
            if (book, j) in dp:
                return dp[(book, j)]

            if book == n:
                return h

            thic, height = books[book]
            cost = float('inf')
            if j + thic <= shelfWidth:
                cost = place(book + 1, j + thic, max(height, h))

            if j != 0:
                cost = min(cost, h + place(book, 0, 0))

            dp[(book, j)] = cost
            return cost

        ans = place(0, 0, 0)
        # print(dp)
        return ans

File: Filling Bookcase Shelves.go
func key(book int, j int) int {
    return book << 10 | j
}

func minHeightShelves(books [][]int, shelfWidth int) int {
    n := len(books)
    dp := make(map[int]int)
    var place func(int, int, int) int
    place = func(book int, j int, h int) int {
        k := key(book, j)
        if val, ok := dp[k]; ok {
            return val
        }

        if book == n {
            return h
        }

        thic, height := books[book][0], books[book][1]
        cost := 1 << 32
        if j + thic <= shelfWidth {
            cost = place(book+1, j+thic, max(height, h))
        }

        if j != 0 {
            cost = min(cost, h+place(book, 0, 0))
        }

        dp[k] = cost
        return cost
    }

    ans := place(0, 0, 0)
    return ans
}
